src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -109735,6 +109735,996 @@ class Program
+            _ = s.ToString();
+        }
+    }
+}";
+            verify(source0);
+            verify(source0.Replace("object.Equals", "object.ReferenceEquals"));
+
+            void verify(string source)
+            {
+                var comp = CreateCompilation(source);
+                comp.VerifyDiagnostics(
+                        // (10,17): warning CS8602: Dereference of a possibly null reference.
+                        //             _ = s.ToString(); // 1
+                        Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(10, 17),
+                        // (23,17): warning CS8602: Dereference of a possibly null reference.
+                        //             _ = s.ToString(); // 2
+                        Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(23, 17));
+            }
+        }
+
+        [Fact, WorkItem(35816, "https://github.com/dotnet/roslyn/issues/35816")]
+        public void StaticEqualsMethod_MaybeNullExpr_ConstantNull_NoWarningWhenFalse()
+        {
+            var source0 = @"
+#nullable enable
+class C
+{
+    static void M1(string? s)
+    {
+        const string? s2 = null;
+        if (object.Equals(s, s2))
+        {
+            _ = s.ToString(); // 1
+        }
+        else
+        {
+            _ = s.ToString();
+        }
+    }
+
+    static void M2(string? s)
+    {
+        const string? s2 = null;
+        if (object.Equals(s2, s))
+        {
+            _ = s.ToString(); // 2
+        }
+        else
+        {
+            _ = s.ToString();
+        }
+    }
+}";
+            verify(source0);
+            verify(source0.Replace("object.Equals", "object.ReferenceEquals"));
+
+            void verify(string source)
+            {
+                var comp = CreateCompilation(source);
+                comp.VerifyDiagnostics(
+                        // (10,17): warning CS8602: Dereference of a possibly null reference.
+                        //             _ = s.ToString(); // 1
+                        Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(10, 17),
+                        // (23,17): warning CS8602: Dereference of a possibly null reference.
+                        //             _ = s.ToString(); // 2
+                        Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(23, 17));
+            }
+        }
+
+        [Fact, WorkItem(35816, "https://github.com/dotnet/roslyn/issues/35816")]
+        public void StaticEqualsMethod_NonNullExpr_MaybeNullExpr_NoWarning()
+        {
+            var source = @"
+#nullable enable
+class C
+{
+    static void M1(string s)
+    {
+        string? s2 = null;
+        if (object.Equals(s, s2))
+        {
+            _ = s.ToString();
+        }
+        else
+        {
+            _ = s.ToString();
+        }
+    }
+
+    static void M2(string s)
+    {
+        string? s2 = null;
+        if (object.Equals(s2, s))
+        {
+            _ = s.ToString();
```

Comment:
This feels wrong. If this line of code was reached, we _will_ throw a NullReferenceException.

Reply:
I understand that this isn't a pure null test, so we can't learn anything from the comparison. But it's a good example of a scenario we could have caught if we differentiated between "This could have null in it" and "This _definitely_ has null in it", and vice versa for not null.

---
In reply to: [298755212](https://github.com/dotnet/roslyn/pull/36722#discussion_r298755212) [](ancestors = 298755212)

Reply:
Right. At the moment we have three categories ranging from highest confidence to lowest:
- a `null` literal or constant
- a not-null state
- a maybe-null state

In this test, we're comparing two expressions: one has a not-null state and the other has a maybe-null state. We're more confident in the not-null state, so we leave it alone.

Reply:
Will mark this resolved, let me know if there's action I should be taking related to this scenario (I do agree that it's questionable)

---
In reply to: [298757714](https://github.com/dotnet/roslyn/pull/36722#discussion_r298757714) [](ancestors = 298757714)

Reply:
No, there's nothing you can do, I'm just grouching.

---
In reply to: [298766107](https://github.com/dotnet/roslyn/pull/36722#discussion_r298766107) [](ancestors = 298766107,298757714)